cpp/src/lance/arrow/arrow_dataset_test.cc

```diff
@@ -58,3 +63,18 @@ TEST_CASE("FileSystemFactory Test") {
   INFO("Expect table: " << table->ToString() << " Actual table: " << actual_table->ToString());
   CHECK(table->Equals(*actual_table));
 }
+
+TEST_CASE("Test CountRows fast path") {
+  auto tmpdir = fs::temp_directory_path();
+  auto path = tmpdir / "test.lance";
+  auto uri = std::string("file://") + path.string();
+
+  auto schema = arrow::schema({arrow::field("key", arrow::utf8())});
+  std::vector<int32_t> values(1000);
+  std::iota(std::begin(values), std::end(values), 1);
+  auto table = ::arrow::Table::Make(schema, {ToArray(values).ValueOrDie()});
+
+  auto dataset = MakeDataset(table, {}, 32).ValueOrDie();
+  auto scanner = dataset->NewScan().ValueOrDie()->Finish().ValueOrDie();
+  CHECK(scanner->CountRows().ValueOrDie() == 1000);
```

Comment:
Anyway to check that the fast core path is hit?

Reply:
I've added some print messages to make sure path is hit.

To have appropriate tests, i guess we can either make some dependency ingestions into the class, or, doing some surgery that writes a file with "correct metadata", but no data? 

Or, set up some code coverage tool in CI?

Should we do that in a follow up ticket?

Reply:
Yeah I'm ok with that, I guess for now just verify by hand if you do count rows on coco it doesn't take forever :)